File: sdk/python/agentfield/harness/providers/opencode.py
"""OpenCode provider using CLI subprocess."""

from __future__ import annotations

import asyncio
import logging
import os
import re
import shutil
import tempfile
import time
from typing import ClassVar, Dict, Optional

from agentfield.harness._cli import (
    estimate_cli_cost,
    extract_final_text,
    resolve_model_and_variant,
    extract_token_usage,
    parse_jsonl,
    run_cli,
    strip_ansi,
)
from agentfield.harness._availability import ensure_cli_available, provider_unavailable
from agentfield.harness._result import FailureType, Metrics, RawResult

logger = logging.getLogger("agentfield.harness.opencode")

# opencode CLI sometimes prints a hard error to stderr but exits 0
# (notably "Model not found", auth errors, schema validation failures).
# These patterns mark stderr as containing a real failure, not just noise
# like the one-time SQLite migration prelude.
_OPENCODE_STDERR_ERROR_PATTERNS = (
    re.compile(r"^Error:", re.MULTILINE),
    re.compile(r"\bModel not found\b"),
    re.compile(r"\bAuthenticationError\b"),
    re.compile(r"\bUnauthorized\b"),
    re.compile(r"\bAPIError\b"),
)


def _prompt_via_stdin() -> bool:
    """Whether to hand the prompt to opencode over stdin instead of argv.

    On Windows the CLI on PATH is usually an npm .cmd shim that runs via
    cmd.exe, whose ~8k command-line cap real prompts blow straight through
    ("The command line is too long."). opencode reads the prompt from stdin
    when the positional arg is absent, so feed it that way there. POSIX keeps
    the battle-tested positional-arg path.
    """
    return os.name == "nt"


def _count_turns_from_events(events: list[dict[str, object]]) -> int:
    """Count opencode turns from JSON events.

    Preferred definition is one turn per ``step_start`` event. If a stream
    has no step markers, fall back to counting ``tool_use`` events.
    """
    step_starts = sum(1 for event in events if event.get("type") == "step_start")
    if step_starts > 0:
        return step_starts

    tool_uses = sum(1 for event in events if event.get("type") == "tool_use")
    return tool_uses


def _cost_from_events(events: list[dict[str, object]]) -> float | None:
    """Sum opencode per-step costs when present in the JSON stream."""
    total_cost = 0.0
    found_cost = False

    for event in events:
        if event.get("type") != "step_finish":
            continue
        part = event.get("part")
        if not isinstance(part, dict):
            continue
        cost = part.get("cost")
        if isinstance(cost, (int, float)) and not isinstance(cost, bool):
            total_cost += float(cost)
            found_cost = True

    return total_cost if found_cost else None


def _tokens_from_events(
    events: list[dict[str, object]],
) -> tuple[dict[str, int], bool]:
    """Sum token counts from opencode ``step_finish.part.tokens`` objects."""
    total = {
        "input_tokens": 0,
        "output_tokens": 0,
        "cache_read_tokens": 0,
        "cache_creation_tokens": 0,
    }
    found_tokens = False

    def _int(value: object) -> int:
        if isinstance(value, bool):
            return 0
        try:
            return int(value)  # type: ignore[arg-type]
        except (TypeError, ValueError):
            return 0

    for event in events:
        if event.get("type") != "step_finish":
            continue
        part = event.get("part")
        if not isinstance(part, dict):
            continue
        tokens = part.get("tokens")
        if not isinstance(tokens, dict):
            continue
        found_tokens = True
        total["input_tokens"] += _int(tokens.get("input"))
        total["output_tokens"] += _int(tokens.get("output")) + _int(
            tokens.get("reasoning")
        )
        cache = tokens.get("cache")
        if isinstance(cache, dict):
            total["cache_read_tokens"] += _int(cache.get("read"))
            total["cache_creation_tokens"] += _int(cache.get("write"))

    return total, found_tokens


def _extract_opencode_event_error(events: list[dict[str, object]]) -> str | None:
    """Pull a meaningful failure message from an in-band JSON error event."""
    for event in events:
        if event.get("type") != "error":
            continue

        for key in ("message", "error", "text"):
            value = event.get(key)
            if isinstance(value, str) and value.strip():
                return value.strip()[:1000]

        part = event.get("part")
        if isinstance(part, dict):
            for key in ("message", "error", "text"):
                value = part.get(key)
                if isinstance(value, str) and value.strip():
                    return value.strip()[:1000]

        return str(event)[:1000]

    return None


def _extract_opencode_error(stderr: str) -> str:
    """Pull the meaningful failure line(s) out of opencode stderr.

    opencode's stderr typically opens with the SQLite migration prelude
    ("Performing one time database migration…") followed by the real error.
    Naively truncating the first 800 chars hides the part that matters,
    so prefer the line carrying the error marker plus a small window of
    context around it.
    """
    lines = stderr.splitlines()
    for i, line in enumerate(lines):
        for pat in _OPENCODE_STDERR_ERROR_PATTERNS:
            if pat.search(line):
                window = lines[max(0, i - 1) : i + 5]
                return "\n".join(window).strip()[:1000]
    return stderr[:1000]


class OpenCodeProvider:
    """OpenCode CLI provider. Invokes ``opencode run`` subprocess (v1.4+)."""

    # Global concurrency limiter: prevents too many simultaneous opencode
    # processes from overwhelming the LLM API with concurrent requests.
    # Each opencode run spawns a full subprocess (pyright, DB migration, etc.)
    # so unbounded concurrency causes rate-limiting and transient failures.
    # Default raised 3 → 10 to match the typical pr-af review fan-out
    # (~6–8 review_dimension phases + 3 meta-lenses); OpenRouter handles
    # this comfortably on Kimi K2.6. Lower via OPENCODE_MAX_CONCURRENT if
    # your provider has tighter per-key rate limits.
    _MAX_CONCURRENT: ClassVar[int] = int(
        os.environ.get("OPENCODE_MAX_CONCURRENT", "10")
    )
    _concurrency_sem: ClassVar[Optional[asyncio.Semaphore]] = None

    # Shared XDG_DATA_HOME across calls when opt-in is enabled. SQLite
    # migrations only run once per process instead of per-call. None means
    # "fresh tempdir per call" (current default).
    _shared_data_dir: ClassVar[Optional[str]] = None

    def __init__(self, bin_path: str = "opencode"):
        self._bin = bin_path

    @classmethod
    def _get_semaphore(cls) -> asyncio.Semaphore:
        if cls._concurrency_sem is None:
            cls._concurrency_sem = asyncio.Semaphore(cls._MAX_CONCURRENT)
        return cls._concurrency_sem

    async def execute(self, prompt: str, options: dict[str, object]) -> RawResult:
        ensure_cli_available("opencode", self._bin)
        sem = self._get_semaphore()
        logger.debug(
            "Waiting for concurrency slot (%d/%d in use)",
            self._MAX_CONCURRENT - sem._value,
            self._MAX_CONCURRENT,
        )
        async with sem:
            return await self._execute_impl(prompt, options)

    async def _execute_impl(self, prompt: str, options: dict[str, object]) -> RawResult:
        # opencode v1.4+ uses the `run` subcommand (replaces deprecated -p/-c flags)
        cmd = [self._bin, "run"]
        cmd.extend(["--format", "json"])

        # --dir is the project root the agent may read and write. project_dir is
        # the canonical field; fall back to cwd when it is unset. Previously cwd
        # took precedence, so a nested cwd under a shared project_dir made
        # opencode treat the rest of the root as an external directory and
        # auto-reject reads/writes of sibling paths — see agentfield#684. This
        # now matches the Go SDK's opencode provider precedence.
        dir_value = options.get("project_dir") or options.get("cwd")
        if isinstance(dir_value, str):
            cmd.extend(["--dir", dir_value])

        # Pass model via -m. A "#variant" suffix on the model (or an explicit
        # options["variant"]) maps to --variant — opencode's provider-specific
        # reasoning effort (e.g. high, max, minimal).
        model_value, variant_value = resolve_model_and_variant(options)
        if model_value:
            cmd.extend(["-m", model_value])
        if variant_value:
            cmd.extend(["--variant", variant_value])

        # opencode v1.14 does not accept --dangerously-skip-permissions on the
        # `run` subcommand — passing it makes yargs print the run-help screen
        # to stdout and exit 0, which the SDK then captures as the LLM
        # response. opencode in non-TTY mode proceeds without permission
        # prompting, so no flag is needed. See agentfield#582.

        # Handle system prompt - prepend to user prompt since OpenCode
        # has no native --system-prompt flag
        effective_prompt = prompt
        system_prompt = options.get("system_prompt")
        if isinstance(system_prompt, str) and system_prompt.strip():
            effective_prompt = (
                f"SYSTEM INSTRUCTIONS:\n{system_prompt.strip()}\n\n"
                f"---\n\nUSER REQUEST:\n{prompt}"
            )

        # Prompt is a positional arg to `opencode run` (not -p) on POSIX; on
        # Windows it goes over stdin instead (see _prompt_via_stdin).
        prompt_via_stdin = _prompt_via_stdin()
        if not prompt_via_stdin:
            cmd.append(effective_prompt)

        env: Dict[str, str] = {}
        env_value = options.get("env")
        if isinstance(env_value, dict):
            env = {
                str(key): str(value)
                for key, value in env_value.items()
                if isinstance(key, str) and isinstance(value, str)
            }

        # Model is passed via -m flag on the run subcommand (see above)

        cwd: Optional[str] = None

        # Per-call XDG_DATA_HOME by default — guarantees session isolation.
        # AGENTFIELD_OPENCODE_REUSE_DATA_DIR=true reuses one dir across calls
        # in this process so SQLite migrations only run once. Opt-in because
        # the implications vary by container layout (read-only /tmp, multi-
        # tenant deployments, etc.) — default behavior is unchanged.
        reuse_data_dir = os.environ.get(
            "AGENTFIELD_OPENCODE_REUSE_DATA_DIR", "false"
        ).strip().lower() in ("1", "true", "yes")

        temp_data_dir: Optional[str] = None
        if reuse_data_dir:
            if type(self)._shared_data_dir is None or not os.path.isdir(
                type(self)._shared_data_dir or ""
            ):
                type(self)._shared_data_dir = tempfile.mkdtemp(
                    prefix=".secaf-opencode-data-shared-"
                )
            data_dir = type(self)._shared_data_dir
        else:
            temp_data_dir = tempfile.mkdtemp(prefix=".secaf-opencode-data-")
            data_dir = temp_data_dir
        env["XDG_DATA_HOME"] = data_dir

        # Wall-clock cap for ONE opencode subprocess. Default 1800s (30min):
        # Kimi K2.6 on a complex review can need 20+ minutes; cutting at 600s
        # (the previous default) was killing slow-but-progressing calls and
        # then re-running them from scratch via the runner's transient retry
        # path. If a call doesn't finish in 30 min, the prompt or the model
        # is wrong — re-running won't help — so we let it fail cleanly.
        # Override via AGENTFIELD_HARNESS_TIMEOUT_SECONDS for tighter caps.
        timeout_seconds = int(
            os.environ.get("AGENTFIELD_HARNESS_TIMEOUT_SECONDS", "1800")
        )

        start_api = time.monotonic()

        try:
            try:
                stdout, stderr, returncode = await run_cli(
                    cmd,
                    env=env,
                    cwd=cwd,
                    timeout=timeout_seconds,
                    input_text=effective_prompt if prompt_via_stdin else None,
                )
            except FileNotFoundError as exc:
                raise provider_unavailable("opencode", self._bin) from exc
            except TimeoutError as exc:
                return RawResult(
                    is_error=True,
                    error_message=str(exc),
                    failure_type=FailureType.TIMEOUT,
                    metrics=Metrics(),
                )
        finally:
            # Only clean up the per-call tempdir; the shared one outlives the
            # call by design (skip the SQLite migration on the next call).
            if temp_data_dir is not None:
                shutil.rmtree(temp_data_dir, ignore_errors=True)

        api_ms = int((time.monotonic() - start_api) * 1000)
        events = parse_jsonl(stdout)
        if events:
            result_text = extract_final_text(events)
        else:
            result_text = stdout.strip() if stdout.strip() else None
        event_error = _extract_opencode_event_error(events)
        clean_stderr = strip_ansi(stderr.strip()) if stderr else ""

        logger.info(
            "opencode finished: returncode=%d stdout=%d chars elapsed=%ds",
            returncode,
            len(stdout),
            api_ms // 1000,
        )
        if not result_text and clean_stderr:
            logger.warning("opencode no stdout. stderr: %s", clean_stderr[:800])

        if returncode < 0:
            failure_type = FailureType.CRASH
            is_error = True
            error_message: str | None = (
                f"Process killed by signal {-returncode}. stderr: {clean_stderr[:500]}"
                if clean_stderr
                else f"Process killed by signal {-returncode}."
            )
        elif returncode != 0 and result_text is None:
            failure_type = FailureType.CRASH
            is_error = True
            error_message = (
                _extract_opencode_error(clean_stderr)
                if clean_stderr
                else (f"Process exited with code {returncode} and produced no output.")
            )
        elif event_error is not None and result_text is None:
            failure_type = FailureType.CRASH
            is_error = True
            error_message = event_error
        elif (
            result_text is None
            and clean_stderr
            and any(pat.search(clean_stderr) for pat in _OPENCODE_STDERR_ERROR_PATTERNS)
        ):
            # opencode sometimes exits 0 even on hard failures like
            # "Model not found" — surface the real error from stderr instead
            # of silently returning empty output that downstream callers
            # interpret as "agent failed to produce a valid result".
            failure_type = FailureType.CRASH
            is_error = True
            error_message = _extract_opencode_error(clean_stderr)
        else:
            failure_type = FailureType.NONE
            is_error = False
            error_message = None

        stream_cost = _cost_from_events(events)
        if stream_cost is not None:
            estimated_cost = stream_cost
        else:
            estimated_cost = estimate_cli_cost(
                model=model_value or "",
                prompt=effective_prompt,
                result_text=result_text,
            )

        num_turns = _count_turns_from_events(events)
        if num_turns == 0 and result_text:
            num_turns = 1

        tokens, found_tokens = _tokens_from_events(events)
        if not found_tokens:
            tokens = extract_token_usage(events)

        return RawResult(
            result=result_text,
            messages=events,
            metrics=Metrics(
                duration_api_ms=api_ms,
                num_turns=num_turns,
                total_cost_usd=estimated_cost,
                session_id="",
                input_tokens=tokens["input_tokens"],
                output_tokens=tokens["output_tokens"],
                cache_read_tokens=tokens["cache_read_tokens"],
                cache_creation_tokens=tokens["cache_creation_tokens"],
                model=model_value,
            ),
            is_error=is_error,
            error_message=error_message,
            failure_type=failure_type,
            returncode=returncode,
        )

File: sdk/go/harness/opencode.go
package harness

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"regexp"
	"runtime"
	"strconv"
	"strings"
	"sync"
	"time"
)

// openCodePromptViaStdin reports whether to hand the prompt to opencode over
// stdin instead of argv. On Windows the CLI on PATH is usually an npm .cmd
// shim that runs via cmd.exe, whose ~8k command-line cap real prompts blow
// straight through ("The command line is too long.") — pr-af style prompts
// that embed diff context die there, and schema retries (which append more
// text) can never recover. opencode reads the prompt from stdin when the
// positional arg is absent, so feed it that way there. POSIX keeps the
// battle-tested positional-arg path. Mirrors the Python SDK's
// _prompt_via_stdin (harness/providers/opencode.py). Package var so tests can
// exercise both paths on any OS.
var openCodePromptViaStdin = runtime.GOOS == "windows"

var (
	openCodeSemaphore chan struct{}
	semOnce           sync.Once
)

const defaultMaxConcurrent = 4

// OpenCodeProvider invokes the opencode CLI as a subprocess.
type OpenCodeProvider struct {
	BinPath   string
	ServerURL string
	runCLI    func(ctx context.Context, cmd []string, env map[string]string, cwd string, timeout int, stdin []byte) (*CLIResult, error)
}

func getSemaphore() chan struct{} {
	semOnce.Do(func() {
		max := defaultMaxConcurrent
		if val := os.Getenv("OPENCODE_MAX_CONCURRENT"); val != "" {
			if i, err := strconv.Atoi(val); err == nil && i > 0 {
				max = i
			}
		}
		openCodeSemaphore = make(chan struct{}, max)
	})
	return openCodeSemaphore
}

// NewOpenCodeProvider creates an OpenCode provider. If binPath is empty,
// it defaults to "opencode".
func NewOpenCodeProvider(binPath, serverURL string) *OpenCodeProvider {
	if binPath == "" {
		binPath = "opencode"
	}
	if serverURL == "" {
		serverURL = os.Getenv("OPENCODE_SERVER")
	}
	return &OpenCodeProvider{BinPath: binPath, ServerURL: serverURL, runCLI: RunCLIWithStdin}
}

func (p *OpenCodeProvider) Execute(ctx context.Context, prompt string, options Options) (*RawResult, error) {
	// opencode 1.14+ moved non-interactive execution to the `run` subcommand.
	// The legacy top-level `-c <dir> -q -p <prompt>` surface was rebound:
	//   -c → --continue (resume previous session)
	//   -p → --password (provider password)
	// so the old invocation made the binary print help and exit without
	// running, leaving callers with empty trajectories. See issue #517.
	//
	// --format json emits a JSONL event stream (step_start / text / step_finish
	// / tool_use / error) instead of plain text, which lets us recover the final
	// message, per-step cost, and turn count, and surface in-band error events.
	cmd := []string{p.BinPath, "run", "--format", "json"}

	// OpenCode uses --dir for the project directory the agent operates on.
	// ProjectDir is the canonical caller-facing field; fall back to Cwd if
	// only that is set so we still honour the caller's explicit working
	// directory.
	dir := options.ProjectDir
	if dir == "" {
		dir = options.Cwd
	}
	if dir != "" {
		cmd = append(cmd, "--dir", dir)
	}

	// Pass model via -m on the run subcommand when supplied. A "#variant"
	// suffix on the model (or an explicit Options.Variant) maps to
	// --variant — opencode's provider-specific reasoning effort (e.g. high,
	// max, minimal).
	modelValue, variantValue := options.resolveModelAndVariant()
	if modelValue != "" {
		cmd = append(cmd, "-m", modelValue)
	}
	if variantValue != "" {
		cmd = append(cmd, "--variant", variantValue)
	}

	// opencode v1.14 does not accept --dangerously-skip-permissions on the
	// `run` subcommand — passing it makes yargs print the run-help screen
	// to stdout and exit 0, which the SDK then captures as the LLM
	// response. opencode in non-TTY mode proceeds without permission
	// prompting, so no flag is needed. See agentfield#582.

	// Prepend system prompt if provided. OpenCode has no native
	// --system-prompt flag, so inline it ahead of the user prompt.
	effectivePrompt := prompt
	if options.SystemPrompt != "" {
		effectivePrompt = fmt.Sprintf(
			"SYSTEM INSTRUCTIONS:\n%s\n\n---\n\nUSER REQUEST:\n%s",
			strings.TrimSpace(options.SystemPrompt), prompt,
		)
	}

	// Prompt is positional on `opencode run` (replaces deprecated -p) on
	// POSIX; on Windows it goes over stdin instead (see openCodePromptViaStdin).
	var stdinPrompt []byte
	if openCodePromptViaStdin {
		stdinPrompt = []byte(effectivePrompt)
	} else {
		cmd = append(cmd, effectivePrompt)
	}

	// Build environment
	env := make(map[string]string)
	for k, v := range options.Env {
		env[k] = v
	}

	// The attribution check keys off the BASE model — a "#variant" suffix
	// must not defeat the openrouter/ prefix match nor leak into the
	// per-model config overlay key.
	if strings.HasPrefix(strings.ToLower(modelValue), "openrouter/") {
		if _, callerSet := env["OPENCODE_CONFIG_CONTENT"]; !callerSet && os.Getenv("OPENCODE_CONFIG_CONTENT") == "" {
			attributionEnv := mergedProcessEnv(env)
			headers := openRouterAttributionHeaders(attributionEnv)
			modelSlug := strings.TrimPrefix(modelValue, "openrouter/")
			if modelSlug != "" && len(headers) > 0 {
				content := map[string]any{
					"provider": map[string]any{
						"openrouter": map[string]any{
							"models": map[string]any{
								modelSlug: map[string]any{"headers": headers},
							},
						},
					},
				}
				if encoded, err := json.Marshal(content); err == nil {
					env["OPENCODE_CONFIG_CONTENT"] = string(encoded)
				}
			}
		}
	}

	sem := getSemaphore()
	select {
	case sem <- struct{}{}:
		defer func() { <-sem }()
	case <-ctx.Done():
		return nil, ctx.Err()
	}

	// Use a temp data dir to isolate opencode state.
	tempDataDir, err := os.MkdirTemp("", ".agentfield-opencode-data-")
	if err != nil {
		return nil, fmt.Errorf("creating temp data dir: %w", err)
	}
	defer os.RemoveAll(tempDataDir)
	env["XDG_DATA_HOME"] = tempDataDir

	startAPI := time.Now()

	cliResult, err := p.runCLI(ctx, cmd, env, options.Cwd, options.timeout(), stdinPrompt)
	apiMS := int(time.Since(startAPI).Milliseconds())

	if err != nil {
		// Check if it's a "not found" error
		if isExecNotFound(err) {
			return &RawResult{
				IsError: true,
				ErrorMessage: fmt.Sprintf(
					"OpenCode binary not found at '%s'. Install OpenCode: https://opencode.ai",
					p.BinPath,
				),
				FailureType: FailureCrash,
				Metrics:     Metrics{},
			}, nil
		}
		// Timeout
		if strings.Contains(err.Error(), "timed out") {
			return &RawResult{
				IsError:      true,
				ErrorMessage: err.Error(),
				FailureType:  FailureTimeout,
				Metrics:      Metrics{DurationAPIMS: apiMS},
			}, nil
		}
		return nil, err
	}

	cleanStderr := StripANSI(strings.TrimSpace(cliResult.Stderr))

	// Parse the JSON event stream. When opencode emitted no parseable events
	// (older versions, or a hard failure before any output), fall back to the
	// trimmed raw stdout so plain-text output is still surfaced.
	events := parseOpenCodeEvents(cliResult.Stdout)
	var resultText string
	if len(events) > 0 {
		resultText = extractOpenCodeFinalText(events)
	} else {
		resultText = strings.TrimSpace(cliResult.Stdout)
	}
	eventError := extractOpenCodeEventError(events)

	raw := &RawResult{
		Result:   resultText,
		Messages: events,
		Metrics: Metrics{
			DurationAPIMS: apiMS,
			SessionID:     "",
		},
		ReturnCode: cliResult.ReturnCode,
	}

	switch {
	case cliResult.ReturnCode < 0:
		raw.FailureType = FailureCrash
		raw.IsError = true
		if cleanStderr != "" {
			raw.ErrorMessage = fmt.Sprintf("Process killed by signal %d. stderr: %.500s",
				-cliResult.ReturnCode, cleanStderr)
		} else {
			raw.ErrorMessage = fmt.Sprintf("Process killed by signal %d.", -cliResult.ReturnCode)
		}
	case cliResult.ReturnCode != 0 && resultText == "":
		raw.FailureType = FailureCrash
		raw.IsError = true
		if cleanStderr != "" {
			raw.ErrorMessage = extractOpenCodeError(cleanStderr)
		} else {
			raw.ErrorMessage = fmt.Sprintf("Process exited with code %d and produced no output.", cliResult.ReturnCode)
		}
	case eventError != "" && resultText == "":
		raw.FailureType = FailureCrash
		raw.IsError = true
		raw.ErrorMessage = eventError
	case resultText == "" && cleanStderr != "" && matchesOpenCodeError(cleanStderr):
		// opencode sometimes exits 0 even on hard failures like "Model not
		// found" or auth errors — surface the real error from stderr instead
		// of silently returning empty output that downstream callers would
		// interpret as "the agent produced no valid result".
		raw.FailureType = FailureCrash
		raw.IsError = true
		raw.ErrorMessage = extractOpenCodeError(cleanStderr)
	}

	// Turn count: prefer the event-derived count, else 1 when a result exists.
	numTurns := countTurnsFromEvents(events)
	if numTurns == 0 && resultText != "" {
		numTurns = 1
	}
	raw.Metrics.NumTurns = numTurns
	raw.Metrics.CostUSD = costFromEvents(events)
	tokens, foundTokens := tokenUsageFromOpenCodeEvents(events)
	if !foundTokens {
		tokens = extractTokenUsage(events)
	}
	raw.Metrics.InputTokens = tokens.inputTokens
	raw.Metrics.OutputTokens = tokens.outputTokens
	raw.Metrics.CacheReadTokens = tokens.cacheReadTokens
	raw.Metrics.CacheCreationTokens = tokens.cacheCreationTokens

	return raw, nil
}

// opencode CLI sometimes prints a hard error to stderr but exits 0 (notably
// "Model not found", auth errors, schema-validation failures). These patterns
// mark stderr as carrying a real failure rather than noise like the one-time
// SQLite migration prelude. Ported from providers/opencode.py:29-35.
var openCodeStderrErrorPatterns = []*regexp.Regexp{
	regexp.MustCompile(`(?m)^Error:`),
	regexp.MustCompile(`\bModel not found\b`),
	regexp.MustCompile(`\bAuthenticationError\b`),
	regexp.MustCompile(`\bUnauthorized\b`),
	regexp.MustCompile(`\bAPIError\b`),
}

func matchesOpenCodeError(stderr string) bool {
	for _, pat := range openCodeStderrErrorPatterns {
		if pat.MatchString(stderr) {
			return true
		}
	}
	return false
}

// parseOpenCodeEvents parses opencode's JSONL event stream, skipping any line
// that is not valid JSON (e.g. interleaved plain-text log lines).
func parseOpenCodeEvents(stdout string) []map[string]any {
	var events []map[string]any
	for _, line := range strings.Split(strings.TrimSpace(stdout), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		var event map[string]any
		if err := json.Unmarshal([]byte(line), &event); err != nil {
			continue
		}
		events = append(events, event)
	}
	return events
}

// extractOpenCodeFinalText reconstructs the final assistant text from the event
// stream. Ported from _cli.extract_final_text (the branches opencode emits):
// step_start resets the accumulated text; "text" events append part/text
// content; result/message/assistant/turn.completed/item.completed carry a final
// message directly.
func extractOpenCodeFinalText(events []map[string]any) string {
	var resultText string
	var currentParts []string

	for _, event := range events {
		eventType, _ := event["type"].(string)
		switch eventType {
		case "step_start":
			currentParts = nil
		case "item.completed":
			if item, ok := event["item"].(map[string]any); ok {
				if it, _ := item["type"].(string); it == "agent_message" {
					if text, ok := item["text"].(string); ok && text != "" {
						resultText = text
					}
				}
			}
		case "result":
			if r, ok := event["result"].(string); ok {
				resultText = r
			} else if r, ok := event["text"].(string); ok {
				resultText = r
			}
		case "turn.completed":
			if text, ok := event["text"].(string); ok && text != "" {
				resultText = text
			}
		case "message", "assistant":
			if content, ok := event["content"].(string); ok && content != "" {
				resultText = content
			} else if content, ok := event["text"].(string); ok && content != "" {
				resultText = content
			}
		case "text":
			content := stringField(event, "text")
			if content == "" {
				content = stringField(event, "content")
			}
			if content == "" {
				if part, ok := event["part"].(map[string]any); ok {
					content = stringField(part, "text")
				}
			}
			if content != "" {
				currentParts = append(currentParts, content)
				resultText = strings.Join(currentParts, "")
			}
		}
	}
	return resultText
}

// stringField returns m[key] when it is a non-empty string, else "".
func stringField(m map[string]any, key string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return ""
}

// countTurnsFromEvents counts opencode turns: one per step_start event, or —
// when the stream has no step markers — one per tool_use event. Ported from
// providers/opencode.py:_count_turns_from_events.
func countTurnsFromEvents(events []map[string]any) int {
	stepStarts := 0
	toolUses := 0
	for _, event := range events {
		switch t, _ := event["type"].(string); t {
		case "step_start":
			stepStarts++
		case "tool_use":
			toolUses++
		}
	}
	if stepStarts > 0 {
		return stepStarts
	}
	return toolUses
}

// costFromEvents sums opencode per-step costs from step_finish events. Returns
// nil when no step carried a cost, so callers distinguish "unknown" from
// "$0.00". Ported from providers/opencode.py:_cost_from_events.
func costFromEvents(events []map[string]any) *float64 {
	total := 0.0
	found := false
	for _, event := range events {
		if t, _ := event["type"].(string); t != "step_finish" {
			continue
		}
		part, ok := event["part"].(map[string]any)
		if !ok {
			continue
		}
		// JSON numbers decode to float64; bool cost values never match here,
		// matching the Python guard against isinstance(cost, bool).
		if cost, ok := part["cost"].(float64); ok {
			total += cost
			found = true
		}
	}
	if !found {
		return nil
	}
	return &total
}

// tokenUsageFromOpenCodeEvents sums token counts from opencode's per-step
// step_finish.part.tokens objects. The boolean reports whether any step carried
// a tokens object, allowing callers to fall back to generic event usage shapes
// when opencode did not report part-level tokens at all.
func tokenUsageFromOpenCodeEvents(events []map[string]any) (tokenUsage, bool) {
	var total tokenUsage
	found := false
	for _, event := range events {
		if t, _ := event["type"].(string); t != "step_finish" {
			continue
		}
		part, ok := event["part"].(map[string]any)
		if !ok {
			continue
		}
		tokens, ok := part["tokens"].(map[string]any)
		if !ok {
			continue
		}
		found = true
		total.inputTokens += intField(tokens, "input")
		total.outputTokens += intField(tokens, "output") + intField(tokens, "reasoning")
		if cache, ok := tokens["cache"].(map[string]any); ok {
			total.cacheReadTokens += intField(cache, "read")
			total.cacheCreationTokens += intField(cache, "write")
		}
	}
	return total, found
}

// extractOpenCodeEventError pulls a meaningful failure message from an in-band
// JSON "error" event. Ported from providers/opencode.py:_extract_opencode_event_error.
func extractOpenCodeEventError(events []map[string]any) string {
	for _, event := range events {
		if t, _ := event["type"].(string); t != "error" {
			continue
		}
		for _, key := range []string{"message", "error", "text"} {
			if v := strings.TrimSpace(stringField(event, key)); v != "" {
				return truncate(v, 1000)
			}
		}
		if part, ok := event["part"].(map[string]any); ok {
			for _, key := range []string{"message", "error", "text"} {
				if v := strings.TrimSpace(stringField(part, key)); v != "" {
					return truncate(v, 1000)
				}
			}
		}
		if b, err := json.Marshal(event); err == nil {
			return truncate(string(b), 1000)
		}
		return ""
	}
	return ""
}

// extractOpenCodeError pulls the meaningful failure line(s) out of opencode
// stderr. opencode's stderr typically opens with the SQLite migration prelude
// followed by the real error, so prefer the line carrying an error marker plus
// a small window of context. Ported from
// providers/opencode.py:_extract_opencode_error.
func extractOpenCodeError(stderr string) string {
	lines := strings.Split(stderr, "\n")
	for i, line := range lines {
		for _, pat := range openCodeStderrErrorPatterns {
			if pat.MatchString(line) {
				start := i - 1
				if start < 0 {
					start = 0
				}
				end := i + 5
				if end > len(lines) {
					end = len(lines)
				}
				window := strings.Join(lines[start:end], "\n")
				return truncate(strings.TrimSpace(window), 1000)
			}
		}
	}
	return truncate(stderr, 1000)
}

func mergedProcessEnv(overrides map[string]string) map[string]string {
	merged := make(map[string]string)
	for _, entry := range os.Environ() {
		key, value, found := strings.Cut(entry, "=")
		if found {
			merged[key] = value
		}
	}
	for k, v := range overrides {
		if v == "" {
			delete(merged, k)
		} else {
			merged[k] = v
		}
	}
	return merged
}
